### ISSUE TITLE
Operation Sessions Extension

### ISSUE THREAD

**Issue by jdinan** [openshmem-org/specification#189]
## Problem

The OpenSHMEM library sees individual operations as they occur.  If the application will perform a series of operations (e.g. a loop of puts), it can be beneficial to inform the runtime.

## Proposed Solution

Allow users to provide a performance hint that they will issue a set of operations.  This information can be used to optimize the behavior of the OpenSHMEM runtime.  For example, sessions may allow the runtime to aggregate operations or completions.

## Proposed API

Straw man API:
```
void shmem_ctx_start_session(shmem_ctx_t ctx);
void shmem_ctx_end_session(shmem_ctx_t ctx);
```

The start session operation informs the OpenSHMEM library that operations will be issued on the given context.  The end session operation informs the OpenSHMEM library that the last operation in the session has been issued on the given context.  Ending a session does not guarantee completion or ordering, quiet/fence on the context must still be used to achieve these guarantees.

**Comment by davidozog** [comment]
Here are the WIP slides from the RMA WG (Min's) discussion today:
https://docs.google.com/presentation/d/1rwONNPFlHMv36eLJSWm8TnGE6nrEvVwpLARaIFH1RmY/edit?usp=sharing

Key takeaways/questions for next time:
1) Do we want bundling to be more "implicit" (the OpenSHMEM library decides how to bundle operations) or "explicit" (the user says exactly which ops are bundled and/or when they are shipped).
2) Should bundles be associated with contexts?
3) Should collectives be supported for bundling?

**Comment by davidozog** [comment]
Here are the slides we covered in the committee meeting today:
https://docs.google.com/presentation/d/1Eco0_yFRS__b-kHy9dM7vO4yDw951_Gc685divy_JZs/edit?usp=sharing

I created a gist that attempts to capture the ideas that came up today.  Feedback is welcome:
https://gist.github.com/davidozog/9be1e52dd425567f86ccb1b050446baf

**Comment by nspark** [comment]
Personally, I like Option 2 from your slides. You present:

```c
shmem_ctx_t my_ctx;
shmem_bundle_start(my_ctx);

shmem_put(my_ctx, ...);
shmem_atomic_inc(my_ctx, ...);
/* ... more RMA/AMO ... */

shmem_bundle_stop(my_ctx);
```

The `shmem_bundle*_t` types seem unnecessary at the current stage of design. I'm hesitant to add structure types and fields if all we're (currently) specifying amounts to a boolean flag.

**Comment by nspark** [comment]
If something more explicit w.r.t. bundle configuration is desired (e.g., to differentiate message aggregation and NIC doorbell chaining), we may want to revisit @naveen-rn's `shmem_ctx_create_ext`, which took a structure-base configuration argument. If the context is the object that "does the bundling," then there shouldn't be a `shmem_bundle_t` type. A sketch:

```c
shmem_ctx_config_t ctx_config = {
  .bundle = SHMEM_CTX_BUNDLE;
};
shmem_ctx_t ctx;
long ret = shmem_ctx_create_ext(SHMEM_CTX_CONFIG, SHMEM_CTX_CONFIG_BUNDLE, &config, &ctx);
if (ret | SHMEM_CTX_CONFIG_BUNDLE)
  printf("Cannot create bundle-enabled context!\n");
else
  printf("Some other error in shmem_ctx_create_ext\n");
```

Maybe, eventually, we'll have something like:

```c
#define SHMEM_CTX_BUNDLE_AGGREGATE (1l << 1)
#define SHMEM_CTX_BUNDLE_CHAIN     (1l << 2)
#define SHMEM_CTX_BUNDLE (SHMEM_CTX_BUNDLE_AGGREGATE | SHMEM_CTX_BUNDLE_CHAIN)
```

**Comment by nspark** [comment]
If bundling itself is a hint, then it seems like the start/stop operations on a bundle-enabled context (I don't like "bundled context") should themselves be hints. An implementation that supports bundling may choose to:

- Only bundle nonblocking operations between explicit start/stop regions
- Bundle nonblocking operations eagerly

**Comment by nspark** [comment]
Re: when to bundle: Should blocking, non-fetching atomics be bundled? They are in many ways like nonblocking operations. Unlike blocking RMA, which may require a copy of the local buffer, these operations take their only payload as a value parameter. I imagine there could be meaningful benefits from NIC doorbell chaining these operations, for example; however, there's no nonblocking, non-fetching AMO operations in the API.

**Comment by naveen-rn** [comment]
> The shmem_bundle*_t types seem unnecessary at the current stage of design.

@nspark I agree with this statement. AFAIU, we have two levels of bundling features - chaining and aggregation. 

Chaining is entirely an one-sided semantics. Meaning, all we need is the source process to group the RMA and AMO events together. To do this, a simple session_start/stop routine is sufficient. I would prefer the following core semantics for chaining:
1. Chaining is an hint for the implementation and performed on a context
2. Chain all NBI RMA, NBI PWS, NBI AMO, and non-fetching blocking AMO operations together
3. Type of chains are implementation defined - meaning chain of a particular set of events (like: NBI RMA-only or NBI-AMO only), or a chain of events to a particular target can be created
4. All chains within a session can be broken explicitly using memory ordering operations like shmem_quiet (or shmem_local_complete and shmem_pe_quiet - if available)
5. Also, session_stop operation implicitly breaks all existing chains within a context
6. Session_stop() doesn't guarantee remote completion (need more discussion) - meaning there is no explicit shmem_quiet() within a session_stop() operation

Aggregation is entirely a different beast to handle at this point. Also, aggregation is sort of a two-sided operation - meaning, the source can aggregate events together and send the aggregated message to the target and it is the responsibility of the target process to unpack the aggregated message. Though, this push-pull sort of semantics can be introduced in SHMEM - we need to discuss more on the usage models.
 
That said, we can easily extend the core chaining-only sessions approach to support aggregation in future.

**Comment by davidozog** [comment]
Here is a use case that came up during the WG call - My apologies, I may have misunderstood/caused confusion at the time (and was being distracted by IMs...) so please correct me if I'm wrong.

``` c
shmem_bundle_start(ctx);
// 1st RMA chain...
shmem_bundle_stop(ctx);   // 1st stop

/* Local work here (we want to overlap) */

shmem_quiet(ctx);

shmem_bundle_start(ctx);
// 2nd RMA chain
shmem_bundle_stop(ctx);  // 2nd stop
```

I think the question is whether a quiet is necessary after the 1st stop, and whether the start/stop bundle API captures the desired optimization.

The quiet is needed if the user wants to assure some overlap is achieved (e.g., if all the start/stop "hints" are removed or are no-ops).  But the runtime can make progress after the 1st stop by breaking the chain (however the implementation does that).  I was thinking (aloud) that a non-blocking quiet could be placed before the local work to enable the optimization without start/stop... but maybe that's slightly off-topic so we ignore that? 😉

Does that capture it @naveen-rn @nspark?  Am I missing anything?

**Comment by nspark** [comment]
I think the quiet is _required_ for this example only if the application itself needs global completion and visibility of the operatons in the first RMA chain. There's no reason that, on its own, your example _needs_ global completion before the start of the second RMA chain. If the application needed that, then the quiet would be required regardless of whether the bundling API was used at all.

To me, this emphasizes that the quiet is _not_ necessary in general. Thus, a program that's structured as "bundle-1 + local work + bundle-2" and is correct (with no quiet), should be correct without bundling. If the quiet is necessary for the program without bundling, then it should be necessary with bundling.

**Comment by nspark** [comment]
> The quiet is needed if the user wants to assure some overlap is achieved (e.g., if all the start/stop "hints" are removed or are no-ops).

In an implementation that _does not_ support bundling, I would expect some natural overlap between the 1st RMA chain and the local compute because of the use of nonblocking operations. As a hint, the bundling helps reduce the NIC doorbell rings; it doesn't change the requirements for overlap.

Put differently, the spec doesn't say a quiet is ever needed for overlap. If anything, it _reduces_ overlap. It shouldn't be necessary in this example unless the application itself required it.

**Comment by davidozog** [comment]
Yup, thanks @nspark - I think we're in agreement.  Yes, my understanding is that this example is presumed to be correct with or without the quiet.  In that case, it's reasonable to expect some overlap between chain 1 and chain 2 without a quiet, but it's not required.   

So a hint-only start/stop API may or may not decide to break the chain after a stop.  A start/stop implementation may decide the user hasn't submitted enough operations to reap any benefit, for example, so it'll wait for the possibility of another chain and make no communication progress.  This would be kind of odd though... I'm just making sure this is what we want before drafting it up.

**Comment by davidozog** [comment]
The RMA WG today discussed an alternative API to `shmem_bundle_start` and `shmem_bundle_stop` that would capture more general hints on a "session".

Here is a possible interface:
```
void shmemx_ctx_session_start(shmem_ctx_t ctx, long options);
void shmemx_ctx_session_stop(shmem_ctx_t ctx);
```

This could support bundling on a context with an option like `SHMEM_CTX_SESSION_BUNDLE`, for example.  Other possible use-cases are a bit less clear, but we thought of a few:

1.  Bundle only certain operations (e.g., only Puts) during a session (`SHMEM_CTX_BUNDLE_PUTS_ONLY`, etc).
2. Indicate to the runtime that contexts will exhibit some sharing property during a session (`SHMEM_CTX_SESSION_PRIVATE_ONLY`, e.g.)
3. Contexts are "read-only" or "write-only" during a session.
4. Contexts can have different priorities or orderings during a session.

Please let me know if I've missed any other use-cases.

It seems this API could retain many of the desired properties of the bundle API as it's proposed in #481 (e.g. no changes to program semantics, the API's are hints-only, etc.)

Do folks have any thoughts or concerns on exploring this direction as an alternatives to the newly proposed `shmem_ctx_bundle` API?

**Comment by jdinan** [comment]
Closed by #493
@davidozog Please start new issues for any follow-on work.
